### ISSUE TITLE
[Feature] New flags for A_SpawnProjectile

### ISSUE THREAD

**Issue by yum13241** [ZDoom/gzdoom#1958]
### GZDoom version

GZDoom 4.10.0

### Which game are you running with GZDoom?

Doom 2

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

EndeavorOS

### Relevant hardware info

N/A

### Is your feature request related to a problem? Please describe.

It's annoying to have to make a duplicate of an actor just to set SeeSound `dsempty`.

This flag would make life easier, and would make code cleaner.

### Describe the solution you'd like

Add CMF_SILENT type flags. I propose
```
CMF_SEESILENT = makes the spawned projectile's SeeSound not play.
CMF_DIESILENT = Idem but for DeathSound.
CMF_BOUNCESILENT = Idem but for BounceSound.
CMF_WALLBOUNCESILENT = Idem but for WallBounceSound.
CMF_TOTALSILENCE  = Idem but no sounds play at all. This includes A_StartSound and A_PlaySound.
```

I might be forgetting something, but you get the idea.

* what does the CMF in flags mean anyway?

### Describe alternatives you've considered

Duplicating actors.

Putting this into a possible A_SpawnProjectileEx?

### Add any other context or screenshots about the feature request here.

Let's take this code. If you run this, the projectile attack sounds loud and bad.

```
Class NewLostSoul : Actor replaces LostSoul
{
	Default
	{
		Health 40;
		Radius 16;
		Height 56;
		Mass 50;
		Speed 8;
		Damage 3;
		PainChance 256;
		Monster;
		+FLOAT +NOGRAVITY +MISSILEMORE +DONTFALL +NOICEDEATH +ZDOOMTRANS +RETARGETAFTERSLAM
		AttackSound "skull/melee";
		PainSound "skull/pain";
		DeathSound "skull/death";
		ActiveSound "skull/active";
		Obituary "$OB_SKULL";
		Tag "$FN_LOST";
	}
	States
	{
	Spawn:
		SKUL AB 10 Bright A_Look();
		Loop;
		
	See:
		SKUL AB 6 Bright A_Chase();
		Loop;
		
	Missile:	
		SKUL C 1 A_Jump(128, "Charge");
		SKUL C 0 A_Jump(128, "Beta-ish");
		Goto Missile;
		
	Charge:
		SKUL C 10 Bright A_FaceTarget();
		SKUL D 4 Bright A_SkullAttack();
		SKUL CD 4 Bright;
		Goto Missile;
		
	Beta-ish:
		SKUL C 10 Bright A_FaceTarget();
		SKUL D 4 Bright
		{
			A_SpawnProjectile("ArachnotronPlasma", 32, 0, 0, 0, 0);
			A_SpawnProjectile("ArachnotronPlasma", 32, -15, 0, 0, 0);
			A_SpawnProjectile("ArachnotronPlasma", 32, 15, 0, 0, 0);
		}
		SKUL CD 4 Bright;
		Goto Missile;
		
	Pain:
		SKUL E 3 Bright;
		SKUL E 3 Bright A_Pain();
		Goto See;
		
	Death:
		SKUL F 6 Bright;
		SKUL G 6 Bright A_Scream();
		SKUL H 6 Bright;
		SKUL I 6 Bright A_NoBlocking();
		SKUL J 6;
		SKUL K 6;
		Stop;
	}
}
```

I *could* do this:

```
Class NewerLostSoul : NewLostSoul replaces LostSoul
{
    States
    {

	Beta-ish:
		SKUL C 10 Bright A_FaceTarget();
		SKUL D 4 Bright
		{
			A_SpawnProjectile("ArachnotronPlasmaSilent", 32, 0, 0, 0, 0);
			A_SpawnProjectile("ArachnotronPlasma", 32, -15, 0, 0, 0);
			A_SpawnProjectile("ArachnotronPlasmaSilent", 32, 15, 0, 0, 0);
		}
		SKUL CD 4 Bright;
		Goto Missile;
    }
}
```
```
Class ArachnotronPlasmaSilent : ArachnotronPlasma
{
    Default
    {
    SeeSound "dsempty";
    }
}
```

LOOK at how much code that took! With my feature in the engine, we could do this:

```
Class NewestLostSoul : Actor replaces LostSoul
{
	Default
	{
		Health 40;
		Radius 16;
		Height 56;
		Mass 50;
		Speed 8;
		Damage 3;
		PainChance 256;
		Monster;
		+FLOAT +NOGRAVITY +MISSILEMORE +DONTFALL +NOICEDEATH +ZDOOMTRANS +RETARGETAFTERSLAM
		AttackSound "skull/melee";
		PainSound "skull/pain";
		DeathSound "skull/death";
		ActiveSound "skull/active";
		Obituary "$OB_SKULL";
		Tag "$FN_LOST";
	}
	States
	{
	Spawn:
		SKUL AB 10 Bright A_Look();
		Loop;
		
	See:
		SKUL AB 6 Bright A_Chase();
		Loop;
		
	Missile:	
		SKUL C 1 A_Jump(128, "Charge");
		SKUL C 0 A_Jump(128, "Beta-ish");
		Goto Missile;
		
	Charge:
		SKUL C 10 Bright A_FaceTarget();
		SKUL D 4 Bright A_SkullAttack();
		SKUL CD 4 Bright;
		Goto Missile;
		
	Beta-ish:
		SKUL C 10 Bright A_FaceTarget();
		SKUL D 4 Bright
		{
			A_SpawnProjectile("ArachnotronPlasma", 32, 0, 0, 0, CMF_SEESILENT);
			A_SpawnProjectile("ArachnotronPlasma", 32, -15, 0, 0, 0);
			A_SpawnProjectile("ArachnotronPlasma", 32, 15, 0, 0, CMF_SEESILENT);
		}
		SKUL CD 4 Bright;
		Goto Missile;
		
	Pain:
		SKUL E 3 Bright;
		SKUL E 3 Bright A_Pain();
		Goto See;
		
	Death:
		SKUL F 6 Bright;
		SKUL G 6 Bright A_Scream();
		SKUL H 6 Bright;
		SKUL I 6 Bright A_NoBlocking();
		SKUL J 6;
		SKUL K 6;
		Stop;
	}
}
```
And that ONLY covers the SeeSound! What if I want a projectile to make noise on sight but not death? Or bounce? It quickly spirals into spaghetti code.

**Comment by Blue-Shadow** [comment]
CMF stands for "Custom Missile Flag", as these flags were originally introduced for the now-deprecated [A_CustomMissile](https://zdoom.org/wiki/A_CustomMissile) function.

**Comment by mc776** [comment]
Not playing the seesound I understand, but what are all the other ones? If they're to change the projectile's post-spawning behaviour each time then I think a new actor class might be cleaner overall.

**Comment by yum13241** [comment]
SeeSound was my main concern. Maybe `CMF_NOALERT` to let it make sound, but not wake stuff up?

**Comment by mc776** [comment]
Alerting monsters is a function of the weapon. For the projectile itself to alert monsters you'd be explicitly calling that in the projectile.

**Comment by Blue-Shadow** [comment]
So the main issue you're having is the loudness of the sound those projectiles make when spawned together at the same time, right? To me, this seems to be a good opportunity to utilize the [$limit](https://zdoom.org/wiki/SNDINFO#limit) command in SNDINFO to solve the problem.

**Comment by yum13241** [comment]
I'm making a monster mod, I don't want to interfere with mods that actually use the sound, for example, as a scripted event. You also can't change it on the fly.

**Comment by Blue-Shadow** [comment]
You could create an [alias](https://zdoom.org/wiki/SNDINFO#alias) to the sound in question, and apply the limit to the alias.

But, if you're really against the idea of creating a new projectile derived from the original with only different sounds, then the above is moot.

**Comment by yum13241** [comment]
> But, if you're really against the idea of creating a new projectile derived from the original with only different sounds, then the above is moot.

Which just reeks of DECORATE to me. (No, I'm not shitting on DECORATE here, it's just that frankly, DECORATE is so limited [ba dum tss] that making duplicates reminds me a bit too much of it in a ZScript monster mod I'm making)

See the OP. 
TL;DR: Why have 2 classes when you could have 1 with my feature?


> Not playing the seesound I understand, but what are all the other ones? If they're to change the projectile's post-spawning behaviour each time then I think a new actor class might be cleaner overall.

I only added the others for completeness's sake.

**Comment by inkoalawetrust** [comment]
I'm not sure how these flags would even work on the projectiles' side of the code. I mean, I guess they could simply remove the specified sound from the projectiles' properties. Like CMF_SEESILENT removing the SeeSound property of an otherwise unmodified projectile actor. But like others have said, this is probably best done with a ZScript wrapper function that uses the pointer to the projectile returned by A_SpawnProjectile to apply modifications immediately after spawning, which will also give you a lot more flexibility as well.

**Comment by yum13241** [comment]
I thought `int`s went up to **2, 147, 483, 647**?

> which will also give you a lot more flexibility as well.

True, but for something simple like what I want you go `LOC` : `Work Done`(1) negative. For something more complex you might as well make a new `Class`.



(1): The ratio of Lines of Code to what the code actually does.

**Comment by inkoalawetrust** [comment]
> I thought `int`s went up to **2, 147, 483, 647**?

Yes, 32-bit ints go up to 2 billion, but bitfields need to use power of two values starting from 1 and going up to -2147483647, which allows multiple flags to be properly checked and assigned. You can also see that in mc776' post where he linked the CMF_ flag field. Though I'm not sure why CMF_AIMMODE was set to 3 and numbers 1 and 2 were skipped.


> True, but for something simple like what I want you go `LOC` : `Work Done`(1) negative. For something more complex you might as well make a new `Class`.
> 
> (1): The ratio of Lines of Code to what the code actually does.

Aren't you already using a new class ? I put together this wrapper for A_SpawnProjectile in about 8 minutes, and it does what you want. Except for being able to make an actor totally silent, which would have to be done with a new actor flag checked for in [VerifyActorSound()](https://sourcegraph.com/github.com/ZDoom/gzdoom@eb94f81a83294024bf87fba8d437042fc07eba7c/-/blob/src/sound/s_doomsound.cpp?L470-495).

```
	Enum SMSFlags
	{
		SMS_SEESILENT 			= 1 << 0, //Missile has no SeeSound when spawned.
		SMS_DIESILENT 			= 1 << 1, //Missile has no DeathSound when spawned.
		SMS_BOUNCESILENT 		= 1 << 2, //Missile has no BounceSound when spawned.
		SMS_WALLBOUNCESILENT 	= 1 << 3, //Missile has no WallBounceSound when spawned.
		SMS_TOTALSILENSE		= SMS_SEESILENT|SMS_DIESILENT|SMS_BOUNCESILENT|SMS_WALLBOUNCESILENT,
	}
	
	Actor SilentMissileSpawn (Class<Actor> Type, Double SpawnHeight = 2, Double SpawnOfs_XY = 0, Double Angle = 0, Int Flags = 0, Int ExtraFlags = 0, Int Ptr = AAPTR_TARGET)
	{
		Actor Missile = A_SpawnProjectile (Type, SpawnHeight, SpawnOfs_XY, Angle, Flags, Pitch, Ptr);
		
		If (!Missile) Return Null; //No missile spawned.
		
		//Silence missile.
		If (ExtraFlags & SMS_SEESILENT) Missile.SeeSound = "";
		If (ExtraFlags & SMS_DIESILENT) Missile.DeathSound = "";
		If (ExtraFlags & SMS_BOUNCESILENT) Missile.BounceSound = "";
		If (ExtraFlags & SMS_WALLBOUNCESILENT) Missile.WallBounceSound = "";
		
		Return Missile;
	}
```

**Comment by yum13241** [comment]
By class I meant projectile class.


> I made a wrapper

Which would be under`extend Class MyVersionOfActor`, right? I'm a ZScript noob, sorry for bothering.


I don't want to sound like a dick, but adding my feature would take up less lines of code and is much simpler. If this gets closed as [DIY] I understand.

**Comment by inkoalawetrust** [comment]
Yes, you can add the function and enum in a class extension or a mixin and use it. As for adding this natively, like mc776 said, only adding a flag for removing the see sound would make sense. Since it would be the most likely to be used. And maybe also a flag to turn on the hypothetical MF9_SILENT flag for the projectile if that is ever added.

**Comment by Blue-Shadow** [comment]
> I put together this wrapper for A_SpawnProjectile in about 8 minutes, and it does what you want.

Note that by the time you get the pointer to the projectile, the SeeSound has already played, so changing it at that point won't achieve the desired result.

**Comment by mc776** [comment]
Calling A_StopSound if SMS_SEESILENT is present would also work, whether in addition to or instead of changing the property.

**Comment by yum13241** [comment]
> Note that by the time you get the pointer to the projectile, the SeeSound has already played, so changing it at that point won't achieve the desired result.

Which is precisely why I can't accept @inkoalawetrust's solution.

> Calling A_StopSound if SMS_SEESILENT is present would also work, whether in addition to or instead of changing the property.

That could be DIY, but at present there's no way to change the SeeSound of a projectile without making a new `Class`.

I don't want my monster mod (or any of my mods for the matter) to become spaghetti code, BRUTAL DOOM style. (ZScript Edition) (TM)

> As for adding this natively, like mc776 said, only adding a flag for removing the see sound would make sense. Since it would be the most likely to be used.

Which makes perfect sense. I only bothered to add the others for completeness anyway.

**Comment by mc776** [comment]
This solution for this particular use case doesn't strike me as particularly spaghetti, though - it's really unusual so there's nothing wrong in having a special dedicated function for it.

Also fwiw if you spawn a projectile using Spawn() the seesound doesn't play.

**Comment by mc776** [comment]
If you're doing this in 50 different places it might take a bit of re-thinking how you're structuring the project, whether or not this one specific flag is available. I literally cannot imagine a mod that would have 50 different projectile shot patterns all of which must use a stock Doom projectile directly, all of which are small enough to require that only one spawn sound be played, and the selection of which cannot be pared down significantly by use of conditions, variables and for loops.

**Comment by yum13241** [comment]
SeeSound only plays on spawn.

**Comment by mc776** [comment]
And bounce for some bounce types.

**Comment by yum13241** [comment]
True.

**Comment by yum13241** [comment]
Also, even mod projectiles could have this problem. I could have a new projectile that plays a `SeeSound`. Duplicating actors is not fun.

**Comment by mc776** [comment]
> True.
> 
> > the plasma, fireball and rocket sounds are all plenty long enough for them to still be playing as the projectile flies past you.
> 
> If they ALL play the sound then it also sounds weird.

\> object that should make a sound, making the sound that it should make, "sounds weird"

i think the pertinent software error is between chair and keyboard on this one

**Comment by yum13241** [comment]
> object that should make a sound, making the sound that it should make, "sounds weird"

I don't think so. They make the sound as it should, but it sounds AWFUL.

Now stop being a jerk.

**Comment by CandiceJoy** [comment]
I think this is unlikely to be approved.

**Comment by coelckers** [comment]
Good call. I totally missed all the fun. Now seriously,  is this all really just about not wanting to make a small inheritance based alteration?
If yes, I'd call this whole thing "Much ado about nothing", i.e. the problem can already be solved with existing features. 
If that's the case asking for the feature anyway won't sway anyone here.

**Comment by yum13241** [comment]
> Now seriously, is this all really just about not wanting to make a small inheritance based alteration?

By that logic we should bring back replacing the options menu. Which we shouldn't.

**Comment by madame-rachelle** [comment]
Yum can you go a single day without a smartass but somehow still totally stupid and irrelevant comment?

P.S. if you ever get a day job, don't quit it for comedy.
